File: problems/tier1-foundation/005-issue-resolver/solution.js
// Issue Resolver — Strategy + Observer reference solution (JavaScript).

const Priority = { LOW: 0, MEDIUM: 1, HIGH: 2, CRITICAL: 3 };

const IssueState = {
  OPEN: "OPEN",
  IN_PROGRESS: "IN_PROGRESS",
  RESOLVED: "RESOLVED",
  CLOSED: "CLOSED",
};

const Category = {
  BILLING: "BILLING",
  TECHNICAL: "TECHNICAL",
  GENERAL: "GENERAL",
  ACCOUNT: "ACCOUNT",
};

function state_name(s) {
  return s;
}

function prio_from_string(s) {
  const m = { LOW: Priority.LOW, MEDIUM: Priority.MEDIUM, HIGH: Priority.HIGH, CRITICAL: Priority.CRITICAL };
  return Object.prototype.hasOwnProperty.call(m, s) ? m[s] : Priority.LOW;
}

class Issue {
  constructor(id, description, category, priority) {
    this.id = id;
    this.description = description;
    this.category = category;
    this.priority = priority;
    this.state = IssueState.OPEN;
    this.assignedAgentId = -1;
  }
}

class Agent {
  constructor(id, name, specialization = null) {
    this.id = id;
    this.name = name;
    this.currentLoad = 0;
    this.specializations = specialization ? [specialization] : [];
  }
}

class AssignmentStrategy {
  selectAgent(agents, issue) { throw new Error('not implemented'); }
}

class RoundRobinStrategy extends AssignmentStrategy {
  constructor() {
    super();
    this.nextIndex = 0;
  }
  selectAgent(agents, issue) {
    if (agents.length === 0) return -1;
    const idx = this.nextIndex % agents.length;
    this.nextIndex = (this.nextIndex + 1) % agents.length;
    return agents[idx].id;
  }
}

class LeastLoadedStrategy extends AssignmentStrategy {
  selectAgent(agents, issue) {
    if (agents.length === 0) return -1;
    let best = agents[0];
    for (let i = 1; i < agents.length; i++) {
      const a = agents[i];
      if (a.currentLoad < best.currentLoad || (a.currentLoad === best.currentLoad && a.id < best.id)) {
        best = a;
      }
    }
    return best.id;
  }
}

class SpecialistStrategy extends AssignmentStrategy {
  constructor() {
    super();
    this.fallback = new LeastLoadedStrategy();
  }
  selectAgent(agents, issue) {
    if (agents.length === 0) return -1;
    const specialists = agents.filter(a => a.specializations.includes(issue.category));
    if (specialists.length === 0) {
      return this.fallback.selectAgent(agents, issue);
    }
    let best = specialists[0];
    for (let i = 1; i < specialists.length; i++) {
      const a = specialists[i];
      if (a.currentLoad < best.currentLoad || (a.currentLoad === best.currentLoad && a.id < best.id)) {
        best = a;
      }
    }
    return best.id;
  }
}

class IssueObserver {
  onStateChange(issueId, oldState, newState) { throw new Error('not implemented'); }
}

class LoggingObserver extends IssueObserver {
  constructor(log) {
    super();
    this.log = log;
  }
  onStateChange(issueId, oldState, newState) {
    this.log.push(`Issue ${issueId}: ${oldState} -> ${newState}`);
  }
}

class IssueResolver {
  constructor(strategy) {
    this.strategy = strategy;
    this.observers = [];
  }
  addObserver(obs) {
    this.observers.push(obs);
  }
  assign(agents, issues, issue) {
    const agentId = this.strategy.selectAgent(agents, issue);
    issue.assignedAgentId = agentId;
    for (const a of agents) {
      if (a.id === agentId) {
        a.currentLoad += 1;
        break;
      }
    }
    issues.push(issue);
    return issue;
  }
  getAgentIssues(issues, agentId) {
    return issues.filter(i => i.assignedAgentId === agentId);
  }
  transitionState(issues, issueId, newState) {
    for (const issue of issues) {
      if (issue.id !== issueId) continue;
      const old = issue.state;
      const valid =
        (old === IssueState.OPEN && newState === IssueState.IN_PROGRESS) ||
        (old === IssueState.IN_PROGRESS && newState === IssueState.RESOLVED) ||
        (old === IssueState.RESOLVED && newState === IssueState.CLOSED);
      if (!valid) return false;
      issue.state = newState;
      for (const obs of this.observers) {
        obs.onStateChange(issueId, old, newState);
      }
      return true;
    }
    return false;
  }
}

let _g_agents = [];
let _g_issues = [];
let _g_issue_id = 0;
let _g_rr = null;
let _g_ir = null;
let _g_log = [];
let _g_logger = null;

function reset_service() {
  _g_agents = [];
  _g_issues = [];
  _g_issue_id = 0;
  _g_rr = new RoundRobinStrategy();
  _g_ir = new IssueResolver(_g_rr);
  _g_log = [];
  _g_logger = new LoggingObserver(_g_log);
  _g_ir.addObserver(_g_logger);
}

function ir_add_agent(id, name, specialization) {
  _g_agents.push(new Agent(id, name, specialization ? specialization : null));
}

function ir_assign_issue_round_robin(description, category, priority) {
  _g_issue_id += 1;
  const issue = new Issue(_g_issue_id, description, category, prio_from_string(priority));
  return _g_ir.assign(_g_agents, _g_issues, issue).assignedAgentId;
}

function ir_assign_issue_least_loaded(description, category, priority) {
  _g_issue_id += 1;
  const issue = new Issue(_g_issue_id, description, category, prio_from_string(priority));
  const r = new IssueResolver(new LeastLoadedStrategy());
  return r.assign(_g_agents, _g_issues, issue).assignedAgentId;
}

function ir_assign_issue_specialist(description, category, priority) {
  _g_issue_id += 1;
  const issue = new Issue(_g_issue_id, description, category, prio_from_string(priority));
  const r = new IssueResolver(new SpecialistStrategy());
  return r.assign(_g_agents, _g_issues, issue).assignedAgentId;
}

function ir_agent_issue_count(agentId) {
  return _g_ir.getAgentIssues(_g_issues, agentId).length;
}

function ir_agent_load(agentId) {
  for (const a of _g_agents) {
    if (a.id === agentId) return a.currentLoad;
  }
  return -1;
}

function ir_transition(issueId, newState) {
  return _g_ir.transitionState(_g_issues, issueId, newState);
}

function ir_get_issue_state(issueId) {
  for (const i of _g_issues) {
    if (i.id === issueId) return i.state;
  }
  return "";
}

function ir_log_size() {
  return _g_log.length;
}

function ir_log_entry(idx) {
  if (idx >= 0 && idx < _g_log.length) return _g_log[idx];
  return "";
}

module.exports = {
  Issue,
  Agent,
  AssignmentStrategy,
  RoundRobinStrategy,
  LeastLoadedStrategy,
  SpecialistStrategy,
  IssueObserver,
  LoggingObserver,
  IssueResolver,
  reset_service,
  ir_add_agent,
  ir_assign_issue_round_robin,
  ir_assign_issue_least_loaded,
  ir_assign_issue_specialist,
  ir_agent_issue_count,
  ir_agent_load,
  ir_transition,
  ir_get_issue_state,
  ir_log_size,
  ir_log_entry,
};

File: problems/tier1-foundation/005-issue-resolver/solution.py
"""Issue Resolver — Strategy + Observer reference solution (Python)."""

from abc import ABC, abstractmethod


class Priority:
    LOW = 0
    MEDIUM = 1
    HIGH = 2
    CRITICAL = 3


class IssueState:
    OPEN = "OPEN"
    IN_PROGRESS = "IN_PROGRESS"
    RESOLVED = "RESOLVED"
    CLOSED = "CLOSED"


class Category:
    BILLING = "BILLING"
    TECHNICAL = "TECHNICAL"
    GENERAL = "GENERAL"
    ACCOUNT = "ACCOUNT"


def state_name(s):
    return s


def prio_from_string(s):
    return {"LOW": Priority.LOW, "MEDIUM": Priority.MEDIUM, "HIGH": Priority.HIGH, "CRITICAL": Priority.CRITICAL}.get(s, Priority.LOW)


class Issue:
    def __init__(self, id, description, category, priority):
        self.id = id
        self.description = description
        self.category = category
        self.priority = priority
        self.state = IssueState.OPEN
        self.assignedAgentId = -1


class Agent:
    def __init__(self, id, name, specialization=None):
        self.id = id
        self.name = name
        self.currentLoad = 0
        self.specializations = [specialization] if specialization else []


class AssignmentStrategy(ABC):
    @abstractmethod
    def selectAgent(self, agents, issue):
        ...


class RoundRobinStrategy(AssignmentStrategy):
    def __init__(self):
        self.nextIndex = 0

    def selectAgent(self, agents, issue):
        if not agents:
            return -1
        idx = self.nextIndex % len(agents)
        self.nextIndex = (self.nextIndex + 1) % len(agents)
        return agents[idx].id


class LeastLoadedStrategy(AssignmentStrategy):
    def selectAgent(self, agents, issue):
        if not agents:
            return -1
        best = agents[0]
        for a in agents[1:]:
            if a.currentLoad < best.currentLoad or (a.currentLoad == best.currentLoad and a.id < best.id):
                best = a
        return best.id


class SpecialistStrategy(AssignmentStrategy):
    def __init__(self):
        self.fallback = LeastLoadedStrategy()

    def selectAgent(self, agents, issue):
        if not agents:
            return -1
        specialists = [a for a in agents if issue.category in a.specializations]
        if not specialists:
            return self.fallback.selectAgent(agents, issue)
        best = specialists[0]
        for a in specialists[1:]:
            if a.currentLoad < best.currentLoad or (a.currentLoad == best.currentLoad and a.id < best.id):
                best = a
        return best.id


class IssueObserver(ABC):
    @abstractmethod
    def onStateChange(self, issueId, oldState, newState):
        ...


class LoggingObserver(IssueObserver):
    def __init__(self, log):
        self.log = log

    def onStateChange(self, issueId, oldState, newState):
        self.log.append(f"Issue {issueId}: {oldState} -> {newState}")


class IssueResolver:
    def __init__(self, strategy):
        self.strategy = strategy
        self.observers = []

    def addObserver(self, obs):
        self.observers.append(obs)

    def assign(self, agents, issues, issue):
        agent_id = self.strategy.selectAgent(agents, issue)
        issue.assignedAgentId = agent_id
        for a in agents:
            if a.id == agent_id:
                a.currentLoad += 1
                break
        issues.append(issue)
        return issue

    def getAgentIssues(self, issues, agent_id):
        return [i for i in issues if i.assignedAgentId == agent_id]

    def transitionState(self, issues, issue_id, new_state):
        for issue in issues:
            if issue.id != issue_id:
                continue
            old = issue.state
            valid = (
                (old == IssueState.OPEN and new_state == IssueState.IN_PROGRESS)
                or (old == IssueState.IN_PROGRESS and new_state == IssueState.RESOLVED)
                or (old == IssueState.RESOLVED and new_state == IssueState.CLOSED)
            )
            if not valid:
                return False
            issue.state = new_state
            for obs in self.observers:
                obs.onStateChange(issue_id, old, new_state)
            return True
        return False


_g_agents = []
_g_issues = []
_g_issue_id = 0
_g_rr = None
_g_ir = None
_g_log = []
_g_logger = None


def reset_service():
    global _g_agents, _g_issues, _g_issue_id, _g_rr, _g_ir, _g_log, _g_logger
    _g_agents = []
    _g_issues = []
    _g_issue_id = 0
    _g_rr = RoundRobinStrategy()
    _g_ir = IssueResolver(_g_rr)
    _g_log = []
    _g_logger = LoggingObserver(_g_log)
    _g_ir.addObserver(_g_logger)


def ir_add_agent(id, name, specialization):
    _g_agents.append(Agent(id, name, specialization if specialization else None))


def ir_assign_issue_round_robin(description, category, priority):
    global _g_issue_id
    _g_issue_id += 1
    issue = Issue(_g_issue_id, description, category, prio_from_string(priority))
    return _g_ir.assign(_g_agents, _g_issues, issue).assignedAgentId


def ir_assign_issue_least_loaded(description, category, priority):
    global _g_issue_id
    _g_issue_id += 1
    issue = Issue(_g_issue_id, description, category, prio_from_string(priority))
    r = IssueResolver(LeastLoadedStrategy())
    return r.assign(_g_agents, _g_issues, issue).assignedAgentId


def ir_assign_issue_specialist(description, category, priority):
    global _g_issue_id
    _g_issue_id += 1
    issue = Issue(_g_issue_id, description, category, prio_from_string(priority))
    r = IssueResolver(SpecialistStrategy())
    return r.assign(_g_agents, _g_issues, issue).assignedAgentId


def ir_agent_issue_count(agent_id):
    return len(_g_ir.getAgentIssues(_g_issues, agent_id))


def ir_agent_load(agent_id):
    for a in _g_agents:
        if a.id == agent_id:
            return a.currentLoad
    return -1


def ir_transition(issue_id, new_state):
    return _g_ir.transitionState(_g_issues, issue_id, new_state)


def ir_get_issue_state(issue_id):
    for i in _g_issues:
        if i.id == issue_id:
            return i.state
    return ""


def ir_log_size():
    return len(_g_log)


def ir_log_entry(idx):
    if 0 <= idx < len(_g_log):
        return _g_log[idx]
    return ""
